Kernel/Storage/NVMe/NVMeController.h

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2021, Pankaj R <pankydev8@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/NonnullRefPtr.h>
+#include <AK/NonnullRefPtrVector.h>
+#include <AK/OwnPtr.h>
+#include <AK/RefPtr.h>
+#include <AK/Tuple.h>
+#include <AK/Types.h>
+#include <Kernel/Bus/PCI/Device.h>
+#include <Kernel/Locking/Spinlock.h>
+#include <Kernel/Memory/TypedMapping.h>
+#include <Kernel/Storage/NVMe/NVMeDefinitions.h>
+#include <Kernel/Storage/NVMe/NVMeNameSpace.h>
+#include <Kernel/Storage/NVMe/NVMeQueue.h>
+#include <Kernel/Storage/StorageController.h>
+
+namespace Kernel {
+
+class NVMeController : public PCI::Device
+    , public StorageController {
+public:
+    static ErrorOr<NonnullRefPtr<NVMeController>> try_initialize(PCI::DeviceIdentifier const&);
+    ErrorOr<void> initialize();
+    explicit NVMeController(PCI::DeviceIdentifier const&);
+    RefPtr<StorageDevice> device(u32 index) const override;
+    size_t devices_count() const override;
+
+protected:
+    bool reset() override;
+    bool shutdown() override;
+    void complete_current_request(AsyncDeviceRequest::RequestResult result) override;
+    template<typename T>
+    void set_ctrl_regs(u32 offset, T val)
+    {
+        *reinterpret_cast<T*>(m_controller_regs.ptr() + offset) = val;
+    }
+    template<typename T>
+    T get_ctrl_regs(u32 offset)
+    {
+        return *reinterpret_cast<T*>(m_controller_regs.ptr() + offset);
+    }
+
+public:
+    bool reset_controller();
+    bool start_controller();
+    u32 get_admin_q_dept();
+    void write64_controller_regs(u32 offset, u64 value) { *reinterpret_cast<u64*>(m_controller_regs.ptr() + offset) = value; };
+
+    u16 submit_admin_command(struct NVMeSubmission& sub, bool sync = false)
+    {
+        // First queue is always the admin queue
+        if (sync) {
+            return m_admin_queue->submit_sync_sqe(sub);
+        }
+        m_admin_queue->submit_sqe(sub);
+        return 0;
+    }
+
+    bool is_admin_queue_ready() { return m_admin_queue_ready; };
+    void set_admin_queue_ready_flag() { m_admin_queue_ready = true; };
+
+private:
+    ErrorOr<void> identify_and_init_namespaces();
+    Tuple<u64, u8> get_ns_features(ByteBuffer& identify_data_struct);
+    ErrorOr<void> create_admin_queue(u8 irq);
+    ErrorOr<void> create_io_queue(u8 irq, u8 qid);
+    void calculate_doorbell_stride()
+    {
+        auto cap = *reinterpret_cast<u64*>(m_controller_regs.ptr() + CC_CAP);
+        m_dbl_stride = (cap >> CAP_DBL_SHIFT) & CAP_DBL_MASK;
+    }
+
+private:
+    PCI::DeviceIdentifier m_pci_device_id;
+    RefPtr<NVMeQueue> m_admin_queue;
+    NonnullRefPtrVector<NVMeQueue> m_queues;
+    NonnullRefPtrVector<NVMeNameSpace> m_namespaces;
+    Memory::TypedMapping<u8> m_controller_regs;
```

Comment:
We could put it this way for this patch and it will be OK for now, but later on we will need to write the structure definition for this mapping instead. This way we don't need to do "ugly" `reinterpret_cast` to access registers with the required bit width access each time, but to rely on the structure definition and the compiler to do this for us.
Whether you want to handle it now or later is completely up to you :)

Reply:
That is actually a good idea! :) I got rid of most of the `reinterpret_cast`. :)
Of course there are some more but I don't think those are required because we need to write those pointers to registers. But eventually we can also see if we can get rid of them :)